make MAX_VOTERS and MAX_CANDIDATES in elections-phragmen configurable. Fix: #11092

sudipghimire533 opened the pull request:
Closes: #11902 
 Polkadot companion: https://github.com/paritytech/polkadot/pull/5815

## Update
- Allow `max_voters` & `max_candidates` to be configurable from `Config` trait of `pallet_elections_phragmen`
- Put value `1000` (for `max_candidates`) & `10_000` (for `max_voters`) while creating runtime from `/bin/node/runtime`. Value taken as was in PR #11790

kianenigma commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -395,9 +392,10 @@ pub mod pallet {
 		) -> DispatchResultWithPostInfo {
 			let who = ensure_signed(origin)?;
 
+			let max_candidate = <T as Config>::MaxCandidates::get();
```
```suggestion
```

kianenigma commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -395,9 +392,10 @@ pub mod pallet {
 			let actual_count = <Candidates<T>>::decode_len().unwrap_or(0) as u32;
 			ensure!(actual_count <= candidate_count, Error::<T>::InvalidWitnessData);
-			ensure!(actual_count <= MAX_CANDIDATES, Error::<T>::TooManyCandidates);
+			ensure!(actual_count <= max_candidate, Error::<T>::TooManyCandidates);
```
```suggestion
			ensure!(actual_count <= <T as Config>::MaxCandidates::get(), Error::<T>::TooManyCandidates);
```

kianenigma commented:
Looks mostly good, need to make companions.

sudipghimire533 commented:
Polkadot companion: https://github.com/paritytech/polkadot/pull/5815

Szegoo commented:
@sudipghimire533 Note that you have to update the ```benchmarking.rs``` file also.

sudipghimire533 commented:
still editing....

sudipghimire533 commented:
Hey mate, can you help me with this:
```
error: expected one of `:`, `;`, `=`, `@`, or `|`, found keyword `in`
   --> frame/elections-phragmen/src/benchmarking.rs:402:9
    |
402 |         let e in max_voter_count() .. max_voter_count() * MAXIMUM_VOTE as u32;
    |               ^^ expected one of `:`, `;`, `=`, `@`, or `|`

error: expected one of `:`, `;`, `=`, `@`, or `|`, found keyword `in`
   --> frame/elections-phragmen/src/benchmarking.rs:437:9
    |
437 |         let e in max_voter_count::<T>() .. max_voter_count::<T>() * MAXIMUM_VOTE as u32;
    |               ^^ expected one of `:`, `;`, `=`, `@`, or `|`

error: could not compile `pallet-elections-phragmen` due to 2 previous errors
```

`max_voter_count` is function which just return `T::MaxVoters::get()`

kianenigma commented:
> @Szegoo maybe you can help me here:
> 
> ```
> error: expected one of `:`, `;`, `=`, `@`, or `|`, found keyword `in`
>    --> frame/elections-phragmen/src/benchmarking.rs:394:9
>     |
> 394 |         let e in T::MaxVoters::get() .. T::MaxVoters * MAXIMUM_VOTE as u32;
>     |               ^^ expected one of `:`, `;`, `=`, `@`, or `|`
> 
> error: expected one of `:`, `;`, `=`, `@`, or `|`, found keyword `in`
>    --> frame/elections-phragmen/src/benchmarking.rs:429:9
>     |
> 429 |         let e in T::MaxVoters::get() .. T::MaxVoters::get() * MAXIMUM_VOTE as u32;
>     |               ^^ expected one of `:`, `;`, `=`, `@`, or `|`
> ```

I think you need to wrap it in a parenthesis: https://github.com/paritytech/substrate/blob/54d6f2cbae12fa314f2bcaf6d57fba585162438f/frame/staking/src/benchmarking.rs#L815

sudipghimire533 commented:
Not working same error and neither can I assign it as variable and re-use. @kianenigma do you think you can help me with benchmarking.rs ?

sudipghimire533 commented:
> > Not working same error and neither can I assign it as variable and re-use. @kianenigma do you think you can help me with benchmarking.rs ?
> 
> I have run into the same issue when working on this, the suggested solution from @kianenigma should work. Could you give me access to your branch, so maybe I manage to get it working?

@Szegoo I have invited you to my fork. Also I tried adding `()` as well. Not good from my side. Let's see

sudipghimire533 commented:
> > > Not working same error and neither can I assign it as variable and re-use. @kianenigma do you think you can help me with benchmarking.rs ?
> > 
> > 
> > I have run into the same issue when working on this, the suggested solution from @kianenigma should work. Could you give me access to your branch, so maybe I manage to get it working?
> 
> @Szegoo I have invited you to my fork. Also I tried adding `()` as well. Not good from my side. Let's see

Ahh you guys meant to only wrap the `get`. Gotcha but looks like benchmark verification failing no?
```
Jul 26 20:05:52.022 ERROR runtime::elections-phragmen: Failed to run election. Number of voters exceeded    
````

Szegoo commented:
@sudipghimire533 The errors are now fixed :)

sudipghimire533 commented:
> @sudipghimire533 The errors are now fixed :)

Indeed, a smart wizard named Szegoo same and fixed everything.. Thanks a lot for helping out mate :)

Szegoo commented:
> > @sudipghimire533 The errors are now fixed :)
> 
> Indeed, a smart wizard named Szegoo same and fixed everything.. Thanks a lot for helping out mate :)

NP, note that you should put the polkadot companion in the PR description so that the companion CI checks pass. It should look like this: polkadot companion: https://github.com/paritytech/polkadot/pull/5815

sudipghimire533 commented:
> > > @sudipghimire533 The errors are now fixed :)
> > 
> > 
> > Indeed, a smart wizard named Szegoo same and fixed everything.. Thanks a lot for helping out mate :)
> 
> NP, note that you should put the polkadot companion in the PR description so that the companion CI checks pass. It should look like this: polkadot companion: [paritytech/polkadot#5815](https://github.com/paritytech/polkadot/pull/5815)

https://github.com/paritytech/polkadot/pull/5815

This looks ok?

Szegoo commented:
> > > > @sudipghimire533 The errors are now fixed :)
> > > 
> > > 
> > > Indeed, a smart wizard named Szegoo same and fixed everything.. Thanks a lot for helping out mate :)
> > 
> > 
> > NP, note that you should put the polkadot companion in the PR description so that the companion CI checks pass. It should look like this: polkadot companion: [paritytech/polkadot#5815](https://github.com/paritytech/polkadot/pull/5815)
> 
> [paritytech/polkadot#5815](https://github.com/paritytech/polkadot/pull/5815)
> 
> This looks ok?

You should put it inside the PR description, but don't put it in a list as you did. Here is an example: https://github.com/paritytech/substrate/pull/11831

Szegoo commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -259,6 +248,14 @@ pub mod pallet {
 		#[pallet::constant]
 		type TermDuration: Get<Self::BlockNumber>;
 
+		/// The maximum number of candidates in a phragmen election.
```
@kianenigma Should the docs for `MaxCandidates` and `MaxVoters` be more detailed than this?

kianenigma replied:
Please add a section warning that the election happens onchain, and these numbers determine how big the election will be. Also note that if the maximum candidates are reached, we don't accept anymore. If max voters are reached, we just ignore voters.

kianenigma commented on frame/elections-phragmen/src/benchmarking.rs:
```diff
@@ -459,7 +459,7 @@ benchmarks! {
 	election_phragmen_v {
 		let v in 4 .. 16;
-		let fixed_c = MAX_CANDIDATES / 10;
+		let fixed_c = T::MaxCandidates::get() / 10;
```
You should be using smaller values for test, similar to the old code. 

I would do that by `let max_voters = if cfg!(test) { TEST_MAX_VOTERS } else { T::MaxVoters::get() } `

kianenigma replied:
what I wrote was bullshit :p

kianenigma reviewed (changes requested):
Needs to be able to set smaller values for testing, otherwise running tests and benchmarks of this pallet would take ages. 

You can try yourself by doing `cargo test -p pallet-elections-phragmen --features runtime-benchmarks`

Szegoo commented:
@sudipghimire533 Are you working on this, or may I finish this?

sudipghimire533 commented:
> @sudipghimire533 Are you working on this, or may I finish this?

I seems quite busy on my office work. You can take over. Thanks

Szegoo commented:
> Needs to be able to set smaller values for testing, otherwise running tests and benchmarks of this pallet would take ages.
> 
> You can try yourself by doing `cargo test -p pallet-elections-phragmen --features runtime-benchmarks`

I don't think that is right. Please correct me if I am wrong, but in the code, the `TEST_MAX_VOTERS` and `TEST_MAX_CANDIDATES` constants are removed, so for the tests and benchmarks the values are used from the `Test` implementation of `Config`. I have put the values in the `Test` configuration to be the same as the deleted test constants. So the benchmarks don't take a long time to execute. The non-test values are specified in the runtime config and don't impact the benchmarks.

kianenigma reviewed (approved):
yes, I was wrong! Please just update some docs and this is good to go.

Szegoo commented:
@kianenigma I have updated the docs.

kianenigma commented:
companion needs an update too.

sudipghimire533 commented:
> companion needs an update too.

Can you mention what needed to me updated

ruseinov reviewed (approved):
looks good

ggwpez commented:
/cmd queue -c bench-bot $ pallet dev pallet_elections_phragmen

kianenigma commented:
bot rebase

kianenigma commented:
bot merge

kianenigma commented:
https://github.com/sudipghimire533 @Szegoo can you both paste your Polkadot address for a tip here?

Szegoo commented:
> https://github.com/sudipghimire533 @Szegoo can you both paste your Polkadot address for a tip here? 

I can't edit the PR description :P

sudipghimire533 commented:
Wohh I get tip as well? Cool. My address: 5CK5Wz9BRU5LaiiGAtnysAsjhTnr9ksh7rL4v9HYN1P6i1Hd

sudipghimire533 commented:
> > https://github.com/sudipghimire533 @Szegoo can you both paste your Polkadot address for a tip here?
> 
> I can't edit the PR description, should I provide it somehow else?

If it need to be in PR description you can put it here and I will put in the Description

sudipghimire533 commented:
Mine is: 1FNfKQFHFLp2Fin8Xqz1KhtZ5nVr4RqCM4Z5SGtv6QctCA9

Szegoo commented:
Polkadot address: 126X27SbhrV19mBFawys3ovkyBS87SGfYwtwa8J2FjHrtbmA

kianenigma commented:
/tip small

kianenigma commented:
/tip small

Szegoo commented:
> /tip small

It doesn't seem like this worked, only the first tip was submitted if I am not mistaken.

sudipghimire533 commented:
On my address as well there we no balance whatsoever. Maybe because mine was freshly created address and it didn't even had existential deposit

Szegoo commented:
> On my address as well there we no balance whatsoever. Maybe because mine was freshly created address and it didn't even had existential deposit

Yours was submitted successfully, you can check it on https://www.dotreasury.com/dot/tips

sudipghimire533 commented:
https://www.dotreasury.com/dot/tips/11627463_0xeed4b3487ce0c15cb0cc66f0293eece4f70001ca6f75951d9b7880d8bce3822a


You can see: `value: 0` 
![image](https://user-images.githubusercontent.com/39845901/185070960-f6ce344f-1b5b-4cef-8d90-c26dafd37dcf.png)

Szegoo commented:
That is because no one has yet declared a tip value, you can check how this works in the `tips` pallet(https://github.com/paritytech/substrate/tree/master/frame/tips/)

sudipghimire533 commented:
> That is because no one has yet declared a tip value, you can check how this works in the `tips` pallet(https://github.com/paritytech/substrate/tree/master/frame/tips/)

Is that something I need to do or parity team?

Szegoo commented:
> > That is because no one has yet declared a tip value, you can check how this works in the `tips` pallet(https://github.com/paritytech/substrate/tree/master/frame/tips/)
> 
> Is that something I need to do or parity team?

You don't have to do anything, you need to wait for the tippers to declare the tip value.

athei commented:
This broke cumulus. How was that merged without a cumulus companion? Companion build broken?
